### PR TITLE
impl: add RestRequest class

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832)

### CODE REVIEW

File: google/cloud/internal/rest_request.cc [hunk 1]
```diff
@@ -0,0 +1,112 @@
+// Copyright 2021 Google LLC
```
Comment (coryan) [inline google/cloud/internal/rest_request.cc:1]: nit: it is now 2022.

File: google/cloud/internal/rest_request.cc [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (coryan) [inline google/cloud/internal/rest_request.cc:7]: nit: we are now using `https://www...`

File: google/cloud/internal/rest_request.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
+// Copyright 2021 Google LLC
```
Comment (coryan) [inline google/cloud/internal/rest_request.h:1]: same nits in the boilerplate.

File: google/cloud/internal/rest_request_test.cc [hunk 4]
```diff
@@ -0,0 +1,144 @@
+// Copyright 2021 Google LLC
```
Comment (coryan) [inline google/cloud/internal/rest_request_test.cc:1]: ditto.

File: google/cloud/internal/rest_request.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+class RestClient;
+
+class RestRequest {
```
Comment (coryan) [inline google/cloud/internal/rest_request.h:30]: Can we have some documentation, even for ourselves, saying that this class is for?  When would I use it?  What does it represent?

File: google/cloud/internal/rest_request.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  std::string const& path() const;
+  HttpHeaders const& headers() const;
+  HttpParameters const& parameters() const;
+
+  RestRequest& SetPath(std::string path);
+  // Adding a header/value pair that already exists results in the new value
+  // appended to the list of values for the existing header.
+  RestRequest& AddHeader(std::string header, std::string value);
+  RestRequest& AddHeader(std::pair<std::string, std::string> header);
+  // Adding a duplicate param and or value results in both the new and original
```
Comment (coryan) [inline google/cloud/internal/rest_request.h:55]: nit: blank line before the next overload set?

File: google/cloud/internal/rest_request.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  RestRequest(std::string path, HttpHeaders headers);
+  RestRequest(std::string path, HttpParameters parameters);
+  RestRequest(std::string path, HttpHeaders headers, HttpParameters parameters);
+
+  RestRequest(RestRequest const&) = default;
+  RestRequest& operator=(RestRequest const&) = default;
+  RestRequest(RestRequest&&) = default;
+  RestRequest& operator=(RestRequest&&) = default;
+
+  std::string const& path() const;
```
Comment (coryan) [inline google/cloud/internal/rest_request.h:46]: Any reason not to inline these functions?  They seem pretty small.

File: google/cloud/internal/rest_request.h [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+class RestClient;
+
+class RestRequest {
+ public:
+  using HttpHeaders = std::unordered_map<std::string, std::vector<std::string>>;
+  using HttpParameters = std::vector<std::pair<std::string, std::string>>;
+
+  RestRequest();
+  explicit RestRequest(std::string path);
+  RestRequest(std::string path, HttpHeaders headers);
```
Comment (coryan) [inline google/cloud/internal/rest_request.h:37]: Given the modifiers to add headers and parameters, do you need these constructors?  Do you think they will be commonly used?

File: google/cloud/internal/rest_request.h [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  RestRequest(std::string path, HttpParameters parameters);
+  RestRequest(std::string path, HttpHeaders headers, HttpParameters parameters);
+
+  RestRequest(RestRequest const&) = default;
```
Comment (coryan) [inline google/cloud/internal/rest_request.h:41]: Do you need to spell these out?  I would have hoped the compiler generated them for you?

File: google/cloud/internal/rest_request.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+  RestRequest& SetPath(std::string path);
+  // Adding a header/value pair that already exists results in the new value
+  // appended to the list of values for the existing header.
+  RestRequest& AddHeader(std::string header, std::string value);
+  RestRequest& AddHeader(std::pair<std::string, std::string> header);
+  // Adding a duplicate param and or value results in both the new and original
+  // pairs stored in order of addition.
+  RestRequest& AddQueryParameter(std::string parameter, std::string value);
+  RestRequest& AddQueryParameter(std::pair<std::string, std::string> parameter);
```
Comment (coryan) [inline google/cloud/internal/rest_request.h:58]: If you are planning to return `RestRequest&` you may want to consider `&&` overloads so `RestRequest().AddQueryParameter(...).AddHeader(...)` returns a rvalue reference.

**Review by coryan — CHANGES_REQUESTED** [review]
A bunch of nits and things for you to consider, but otherwise looks great.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832)* status: 0 of 6 files reviewed, 10 unresolved discussions (waiting on @coryan and @scotthart)

___
*[google/cloud/internal/rest_request.cc, line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832#-MslFUZPzGi4831VerpD-r1-1:-MslfpT2Cgeanf8RUvRz:b2cctl) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6bea6b6b5ca3f911f22cadd907361e64af93808d/google/cloud/internal/rest_request.cc#L1)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: it is now 2022.
</blockquote></details>

So these files existed in a public branch on my fork back in 2021. I wasn't sure whether or not to bump them to 2022.

**Comment by coryan** [comment]
___
*[google/cloud/internal/rest_request.cc, line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832#-MslFUZPzGi4831VerpD-r1-1:-Msll5IsD_ojBuc_UvFt:b-5wqdgg) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6bea6b6b5ca3f911f22cadd907361e64af93808d/google/cloud/internal/rest_request.cc#L1)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

So these files existed in a public branch on my fork back in 2021. I wasn't sure whether or not to bump them to 2022.
</blockquote></details>

I am not a lawyer, nor do I play one on TV.  My intuition is that we should label them based on the date the file gets merged into the repo, but that is just an intuition.  Your call.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832)* status: 0 of 6 files reviewed, 7 unresolved discussions (waiting on @coryan)

___
*[google/cloud/internal/rest_request.h, line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832#-MslFUZPzGi4831VerpE-r1-1:-MspiaHpAYUlqlCjDQpy:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6bea6b6b5ca3f911f22cadd907361e64af93808d/google/cloud/internal/rest_request.h#L1)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

same nits in the boilerplate.
</blockquote></details>

Done.
___
*[google/cloud/internal/rest_request.h, line 30 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832#-MslFUZPzGi4831VerpE-r1-30:-Mspi_iwBV6Xfb3S43Uh:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6bea6b6b5ca3f911f22cadd907361e64af93808d/google/cloud/internal/rest_request.h#L30)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Can we have some documentation, even for ourselves, saying that this class is for?  When would I use it?  What does it represent?
</blockquote></details>

Done.
___
*[google/cloud/internal/rest_request.h, line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832#-MslFUZPzGi4831VerpE-r1-37:-Mspi_3K3EBF3b8eh48z:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6bea6b6b5ca3f911f22cadd907361e64af93808d/google/cloud/internal/rest_request.h#L37)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Given the modifiers to add headers and parameters, do you need these constructors?  Do you think they will be commonly used?
</blockquote></details>

Done.
___
*[google/cloud/internal/rest_request.h, line 41 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832#-MslFUZPzGi4831VerpE-r1-41:-MspiVk98VTDwYGvN_3M:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6bea6b6b5ca3f911f22cadd907361e64af93808d/google/cloud/internal/rest_request.h#L41)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Do you need to spell these out?  I would have hoped the compiler generated them for you?
</blockquote></details>

Done.
___
*[google/cloud/internal/rest_request.h, line 46 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832#-MslFUZPzGi4831VerpE-r1-46:-MspiVRTB3IOuRxPjlB1:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6bea6b6b5ca3f911f22cadd907361e64af93808d/google/cloud/internal/rest_request.h#L46)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Any reason not to inline these functions?  They seem pretty small.
</blockquote></details>

Done.
___
*[google/cloud/internal/rest_request.h, line 58 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832#-MslFUZPzGi4831VerpE-r1-58:-MspiUSw8x5E8dVvRr1R:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6bea6b6b5ca3f911f22cadd907361e64af93808d/google/cloud/internal/rest_request.h#L58)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

If you are planning to return `RestRequest&` you may want to consider `&&` overloads so `RestRequest().AddQueryParameter(...).AddHeader(...)` returns a rvalue reference.
</blockquote></details>

Done.
___
*[google/cloud/internal/rest_request_test.cc, line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832#-MslFUZPzGi4831VerpF-r1-1:-MspiShu2Hp4w2oVspWD:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6bea6b6b5ca3f911f22cadd907361e64af93808d/google/cloud/internal/rest_request_test.cc#L1)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

ditto.
</blockquote></details>

Done.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832)* status: 0 of 6 files reviewed, 7 unresolved discussions (waiting on @coryan)

___
*[google/cloud/internal/rest_request.h, line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832#-MslFUZPzGi4831VerpE-r1-37:-Mspj9H7DiO-_66bImy_:bf6g727) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6bea6b6b5ca3f911f22cadd907361e64af93808d/google/cloud/internal/rest_request.h#L37)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Commonly? I'm not certain, but if you have a collection of these its easier to call one of these constructors rather than write some loop.

**Review by coryan — APPROVED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7832)* status: 3 of 6 files reviewed, all discussions resolved (waiting on @coryan)

**Pull request merged**